proposals/0001-dependency-injection.md

```diff
@@ -0,0 +1,574 @@
+using configuration.
+
+For example, when using YAML configuration: 
+
+```yaml
+# application.yaml / ktor / application 
+modules:
+   - io.ktor.example.ImplementationKt.declareDependencies
+   - io.ktor.example.ApplicationKt.configureApplication
+```
+
+This way, we can introduce a dependency inversion where the `declareDependencies` module can live on a different runtime
+classpath than the application module `configureApplication`, which has no knowledge of the implementation details.
+
+In the next two sections, we'll provide the API design for both declaring and resolving dependencies.
+
+## Declaration
+[declaration]: #declaration
+
+As with declaring extensions in a Ktor application, users can expect to declare their dependencies from the
+same scope: `Application`.
+
+Here is how to declare dependencies from an application module:
+
+```kotlin
+fun Application.declareDependencies() {
+    dependencies {
+        // Base case
+        provide<DataSource> { PostgresDataSource("jdbc:postgresql://localhost:5440/test") }
+        // With named instance, using resolve() from the provider context
+        provide<DataSource>("mongo") { MongoDataSource(resolve()) }
+        // Using constructor injection
+        provide<Repository<Message>>(MessageRepository::class)
+        // Alternatively, without the optional interface, and using the context 
+        provide {
+            create(::RoomRepository)
+        }
+    }
+}
+```
+
+### Naming
+[naming]: #naming
+
+For the *dependencies* scope:
+
+| Alternative | Reason for avoiding                  |
+|-------------|--------------------------------------|
+| di          | Not immediately recognizable acronym |
+| inject      | confusing for Koin users             |
+| providers   | too abstract                         |
+
+For the *provide* function:
+
+| Alternative | Reason for avoiding                                    |
+|-------------|--------------------------------------------------------|
+| inject      | confusing for Koin users                               |
+| supply      | this suggests that there is some scarcity of instances |
+| declare     | too abstract and non-operative                         |
+| single      | awkward wording out of context                         |
+
+For the *create* function:
+
+| Alternative | Reason for avoiding                                          |
+|-------------|--------------------------------------------------------------|
+| construct   | too specific; this call should allow non-constructor lambdas |
+| inject      | not the most important detail of the function                |
+
+
+## Resolution
+[resolution]: #resolution
+
+In practice, we can expect the declaring modules to be few, and the resolving modules to be many.
+
+Here are the examples of how to resolve dependencies from a Ktor module:
+
+```kotlin
+fun Application.configureApplication() {
```

Comment:
Could you tell me what you think about supporting parameters injection for the modules declared in the config file?

```config.yml
dependencies:
    mongo: DatabaseKt.mongoProduction

modules:
   - ModuleKt.module
   
```

```Module.kt
fun Application.module(mongo: DataSource) {
    // ...
}
```

Reply:
We also can allow registries from DI framework to contribute to shared Ktor registry, so they can benefit from the syntax we have

Reply:
Yep, we can make the Environment implementation pluggable